### PR TITLE
Bug 1941334: Add support for auto pinning policy

### PR DESCRIPTION
**Pull request by liranr23**
This commit will add the required support for the auto pinning feature.
The feature will allow VMs to have their CPU topology, CPU pinning and
NUMA pinning to be automated based on the host hardware upon creation.

There are three optional policies:
- `adjust` policy will change the CPU topology based on the host
  topology, will set the vCPU pinning and NUMA nodes.
- `existing` policy will set the vCPU pinning and NUMA nodes with the
  without overriding the current VM CPU topology.
- `disabled` won't do nothing. It will act like nothing provided.

Each VM using a policy will be pinned to each host within the cluster
with allowed migration.

- This feature is available using 4.4.5 engine version or above. In case
  we are using a lower version, a warning will be shown and the policy
  won't take any action.
- This patch includes a bump of the go-ovirt SDK.

### CODE REVIEW

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 1]
```diff
@@ -120,6 +120,41 @@ func (is *InstanceService) InstanceCreate(
 		}
 	}
 
+	isAutoPinning := false
+	// TODO: remove the version check when everyone uses engine 4.4.5
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:124]: Can this be exported to a function?
L125-134

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 2]
```diff
@@ -120,6 +120,41 @@ func (is *InstanceService) InstanceCreate(
...
+	versionCompareResult, err := versionCompare(engineVer, ovirtsdk.NewVersionBuilder().
+		Major(4).
+		Minor(4).
+		Build_(5).
+		Revision(0).
+		MustBuild())
+	if err != nil {
+		return nil, err
+	}
+	if providerSpec.AutoPinningPolicy != "" && versionCompareResult < 0 {
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:135]:
```
if providerSpec.AutoPinningPolicy != "" {
  versionCompareResult, err := engineVersionGreaterThen(autoPiningRequiredEngineVersion)
...
```

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 3]
```diff
@@ -120,6 +120,41 @@ func (is *InstanceService) InstanceCreate(
...
+		klog.Warning("The engine version %d.%d.%d is not supporting the auto pinning feature. " +
+		"Please update to 4.4.5 or later.", engineVer.MustMajor(), engineVer.MustMinor(), engineVer.MustBuild())
+	} else if providerSpec.AutoPinningPolicy != "" {
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:138]: Guess this else if will be removed once the version check will be inside the autoPining condition

Comment (liranr23) [reply pkg/cloud/ovirt/clients/machineservice.go:138]: and left only with `else`

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 4]
```diff
@@ -423,3 +458,77 @@ func (is *InstanceService) handleAffinityGroups(vm *ovirtsdk.Vm, cID string, ags
...
+	versionCompareResult, err := versionCompare(engineVer, autoPiningRequiredEngineVersion)
+	if err != nil {
+		return false, err
+	}
+	if versionCompareResult >= 0 {
+		return true, nil
+	}
+	klog.Warning("The engine version %d.%d.%d is not supporting the auto pinning feature. " +
+		"Please update to 4.4.5 or later.", engineVer.MustMajor(), engineVer.MustMinor(), engineVer.MustBuild())
+	return false, nil
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:533]: I think you should return an error and not a warning in the logs here,
In the current case things will not happen and there will be some hidden log in a place no one is looking at...
After https://github.com/openshift/cluster-api-provider-ovirt/pull/96 we should port that part of the version check to the validateMachine function and return an InvalidMachineConfiguration error

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 5]
```diff
@@ -423,3 +458,77 @@ func (is *InstanceService) handleAffinityGroups(vm *ovirtsdk.Vm, cID string, ags
...
+	if err != nil {
+		return nil, errors.Errorf("failed to get the list of hosts in the cluster: %v", err)
+	}
+	return hostsInCluster.MustHosts(), nil
+}
+
+// handleAutoPinning updates the VM after creation to set the auto pinning policy configuration.
+func (is *InstanceService) handleAutoPinning(id string, autoPinningPolicy ovirtsdk.AutoPinningPolicy) error {
+	vmService := is.Connection.SystemService().VmsService().VmService(id)
+	optimizeCpuSettings := !(autoPinningPolicy == ovirtsdk.AUTOPINNINGPOLICY_EXISTING)
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:481]: (autoPinningPolicy != ovirtsdk.AUTOPINNINGPOLICY_EXISTING)

So optimizeCpuSettings should be for disabled and adjust?

File: pkg/cloud/ovirt/clients/machineservice.go (hunk 4 above)
Comment (Gal-Zaidman) [reply pkg/cloud/ovirt/clients/machineservice.go:533]: The PR got merges so we should move this check to the validations

File: pkg/cloud/ovirt/clients/machineservice.go (hunk 5 above)
Comment (liranr23) [reply pkg/cloud/ovirt/clients/machineservice.go:481]: No, we did missed the disabled option by using it. When we call the service(API) in this case it is only for existing/adjust. When you have disabled, just don't do anything. I did outside the function now.

File: pkg/cloud/ovirt/clients/machineservice.go [hunk 6]
```diff
@@ -120,6 +120,30 @@ func (is *InstanceService) InstanceCreate(
 		}
 	}
 
+	isAutoPinning := false
+	if providerSpec.AutoPinningPolicy != "" {
```
Comment (Gal-Zaidman) [inline pkg/cloud/ovirt/clients/machineservice.go:124]: why not:
providerSpec.AutoPinningPolicy != "" && autoPinningPolicy != ovirtsdk.AUTOPINNINGPOLICY_DISABLED

File: pkg/cloud/ovirt/clients/machineservice.go (hunk 5 above)
Comment (Gal-Zaidman) [reply pkg/cloud/ovirt/clients/machineservice.go:477]: So I don't understand why are you using:
optimizeCpuSettings := !(autoPinningPolicy == ovirtsdk.AUTOPINNINGPOLICY_EXISTING)

that means that if autoPinningPolicy is adjust then optimizeCpuSettings will be false

File: pkg/cloud/ovirt/clients/machineservice.go (hunk 6 above)
Comment (liranr23) [reply pkg/cloud/ovirt/clients/machineservice.go:124]: as we talked offline, better to leave it as is and change the `isAutoPinning = true` placement.

File: pkg/cloud/ovirt/clients/machineservice.go (hunk 5 above)
Comment (liranr23) [reply pkg/cloud/ovirt/clients/machineservice.go]: optimizeCpuSettings := autoPinningPolicy == ovirtsdk.AUTOPINNINGPOLICY_ADJUST
will solve it.
If we are in ADJUST then -> true
if we are having something else(only existing is possible) -> false

**Pull request merged**